### PR TITLE
Add new provider: BJ-Share (Brazillian)

### PR DESCRIPTION
**Pull request by DarkSupremo**
- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

Adding new provider, it's an private Brazillian tracker.

### CODE REVIEW

**Comment by labrys** [comment]
Just have to laugh at the name.

**Comment by DarkSupremo** [comment]
Anyone can review it?

File: medusa/providers/torrent/html/bjshare.py [hunk 1]
```diff
@@ -81,6 +81,10 @@ def search(self, search_strings, age=0, ep_obj=None, **kwargs):
         if not self.login():
             return results
 
+        manual_search = "manual_search" in kwargs and kwargs['manual_search']
```
Comment (labrys) [inline medusa/providers/torrent/html/bjshare.py]: kwargs.get('manual_search')

Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Actually, i did that to avoid the variable being None, but in this case, it really does not matter... so changed it to simplify the code

**Comment by DarkSupremo** [comment]
The torrent provider have an setting to group torrents, for now it will only work with that setting disabled, i'll try to implement support for it soon

File: medusa/providers/torrent/html/bjshare.py [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+        # Miscellaneous Options
+        self.supports_absolute_numbering = True
+        self.max_back_pages = 2
+        self.units = ["B", "KB", "MB", "GB", "TB", "PB"]
+
+        # Proper Strings
+        self.proper_strings = ["PROPER", "REPACK", "REAL", "RERIP"]
+
+        # Cache
+        self.cache = tv.Cache(self, min_time=1)
```
Comment (duramato) [inline medusa/providers/torrent/html/bjshare.py]: `min_time=1`

File: medusa/providers/torrent/html/bjshare.py [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+
+        self.quality = {
+            'Full HD': '1080p',
+            'HD': '720p',
+            'BRRip': 'BR-Rip',
+            'Remux': 'BR-Rip',
```
Comment (duramato) [inline medusa/providers/torrent/html/bjshare.py]: `'Remux': 'BR-Rip',` This is incorrect. All but `Full HD` and `HD` should be left intact, the name parser _guesses_ them just fine.

**Comment by duramato** [comment]
This implemetation just feels too hacky. You don't need to construct new release names, you can use the ones the site provides and Medusa will detect them just fine, for example:
This is a result I found from screenshoots avaiable online of the site: `S02 AVI / XviD / MP3 / DVDRip / Legendado / Inglês / Free` for `The simpsons`

image


```
guessit returns the following: `guessit "S02 AVI / XviD / MP3 / DVDRip / Legendado / Inglês / Free"
For: S02 AVI / XviD / MP3 / DVDRip / Legendado / Inglês / Free
GuessIt found: {
    "season": 2,
    "container": "AVI",
    "video_codec": "XviD",
    "audio_codec": "MP3",
    "format": "DVD",
    "title": "Legendado",
    "type": "episode"
}
```
And from there it's just missing the show name which you can add easily, making it:
```
guessit "The Simpsons S02 AVI / XviD / MP3 / DVDRip / Legendado / Inglês / Free"
For: The Simpsons S02 AVI / XviD / MP3 / DVDRip / Legendado / Inglês / Free
GuessIt found: {
    "title": "The Simpsons",
    "season": 2,
    "container": "AVI",
    "video_codec": "XviD",
    "audio_codec": "MP3",
    "format": "DVD",
    "release_group": "Legendado",
    "type": "episode"
}
```

**Comment by DarkSupremo** [comment]
@p0psicles the page layout they use for 'latest 24h releases' is completly different from the normal search layout, that why i did an different parser, i think it's best than an big if clause

(Latest 24h releases page)
image

(Normal search page, without grouping, currently supported)
image

(Search page with grouping enabled, currently not supported)
image


@duramato i'll give it an try sending the name intact, the reason i did this way is because i used it in sickrage in the past, and they had trouble to recognize the name..

Other problem that i have is because they always have the year in the release (some times wrong), and actually i should only put it on few tv shows (Doctor Who, and few others 'remakes'), otherwise medusa will fail to recognize it.

I can't search for an specific episode or season on this tracker too...

**Comment by p0psicles** [comment]
You need to merge develop into your branch

**Comment by duramato** [comment]
@DarkSupremo can't you just browse the torrents page without a query?  It seems possible from the second screenshot you showed. 
If they have a year in the name that's a task for scene exceptions not hacks in the provider. As for the name parsing, sickrage is regex based and only supports scene release names while Medusa's can handle almost every release name as it uses guessit. The provider can't be merged with all this custom code, or as i like to out it, hacks.
On a side note, if you an invite I'm more than happy to take a look at implementing it.

**Comment by DarkSupremo** [comment]
Yes, it was implemented with blank search before, but it's limited to 50
releases only (or multiple requests for pagination) i thought it would be
better this way, that list all torrents from the past 24h in one go...

About the scene exception, i don't think it will be nice to the user having
to add it to every show in the catalog... So what i did is just display the
show release name if the year is included in the search term, you think
it's bad? Or i could just not include year at all, but that could break
some shows...

About the rebuilding the release name, ill remove that to see if it work
fine, the only problem will be the past 24h pages, or i can just rely on
blank searches limited to 50 episodes only

Em sex, 18 de mai de 2018 09:58, duramato <notifications@github.com>
escreveu:

**Comment by duramato** [comment]
50 results is more than enough. Change RSS to use a blank query, and you ~~kill~~ fix 2 birds with one stone. Scene exceptions are required to be manually input for some cases already, especially for localized releases with localized show titles, so it won't be an issue.
Don't include the year in the show name.

**Comment by duramato** [comment]
You can send me the invite to the email in my profile.
Sure

File: medusa/providers/torrent/html/bjshare.py [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+                        # Season with episode included
+                        torrent_name = u"{0} S{1}E{2} {3} {4} {5} {6}" \
+                            .format(title, season, episode, resolution, source, codec, audio)
+                    else:
+                        # Probably movie or something else that does not have season or episode
+                        torrent_name = u"{0} {1} {2} {3} {4}" \
+                            .format(title, resolution, source, codec, audio)
+                    # Removing accents due to an bug that i noticed that the torrent is not being sent to
+                    # qBittorrent when accents are present, there was no error in log.
+                    torrent_name = re.sub('\s+', ' ', self._remove_accents(torrent_name)).strip()
```
Comment (duramato) [inline medusa/providers/torrent/html/bjshare.py]: Accents shouldn't removed, the string should left intact. All it should be sending to qBittorrent is the torrent url if I'm not mistaken, so this may sound like a locale config.

Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: So we need to figure out why it was not sending to torrent client when accent was present.. I'll do some more tests

Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Tested again, and i receive the telegram message just fine saying about the download, but its not added to qbittorrent..

That's the response from qbittorrent when i try to send an torrent with accent:

```2018-05-18 13:04:06 INFO     SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] Beginning to manual snatch release: Black.Clover.-.S01E32.MKV./.H.264./.AAC./.HDTV./.Legendado./.Japonês./.1080p
2018-05-18 13:04:06 INFO     SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] Downloading Black.Clover.-.S01E32.MKV./.H.264./.AAC./.HDTV./.Legendado./.Japonês./.1080p with 8 seeders and 0 leechers and size 547.08 MB from BJ-Share
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] GET URL: https://bj-share.me/torrents.php?action=download&id=153167&authkey=a04a6d622a11a69cacc0bc6180c14c13&torrent_pass=n8y6e8lkljrp78gixy4651vutqueo2pg [Status: 200]
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] User-Agent: Medusa/0.2.2 (Windows; 10; cdb05a61-5ab4-11e8-92f1-7cdd9083bde4)
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] Calling qbittorrent Client
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] GET URL: http://localhost:9496/version/api [Status: 200]
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] User-Agent: Medusa/0.2.2 (Windows; 10; cdb05a61-5ab4-11e8-92f1-7cdd9083bde4)
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] POST URL: http://localhost:9496/login [Status: 200]
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] User-Agent: Medusa/0.2.2 (Windows; 10; cdb05a61-5ab4-11e8-92f1-7cdd9083bde4)
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] With post data: username=**********&password=**********
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] qbittorrent: Requested a POST connection to http://localhost:9496/command/upload with params: None Data: None
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] POST URL: http://localhost:9496/command/upload [Status: 200]
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] User-Agent: Medusa/0.2.2 (Windows; 10; cdb05a61-5ab4-11e8-92f1-7cdd9083bde4)
 Content-Disposition: form-data; name="torrents"; filename*=utf-...
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] qbittorrent: Response to POST request is 
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] GET URL: http://localhost:9496/version/api [Status: 200]
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] User-Agent: Medusa/0.2.2 (Windows; 10; cdb05a61-5ab4-11e8-92f1-7cdd9083bde4)
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] qbittorrent: Requested a POST connection to http://localhost:9496/version/api with params: None Data: {u'hash': '2b44e208ac311fae598013251bb7bfbd9e16ba9d'}
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] POST URL: http://localhost:9496/version/api [Status: 200]
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] User-Agent: Medusa/0.2.2 (Windows; 10; cdb05a61-5ab4-11e8-92f1-7cdd9083bde4)
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] With post data: hash=2b44e208ac311fae598013251bb7bfbd9e16ba9d
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] qbittorrent: Response to POST request is 17
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] GET URL: http://localhost:9496/version/api [Status: 200]
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] User-Agent: Medusa/0.2.2 (Windows; 10; cdb05a61-5ab4-11e8-92f1-7cdd9083bde4)
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] GET URL: http://localhost:9496/version/api [Status: 200]
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] User-Agent: Medusa/0.2.2 (Windows; 10; cdb05a61-5ab4-11e8-92f1-7cdd9083bde4)
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] qbittorrent: Requested a POST connection to http://localhost:9496/command/setCategory with params: None Data: {u'category': u'medusa', u'hashes': '2b44e208ac311fae598013251bb7bfbd9e16ba9d'}
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] POST URL: http://localhost:9496/command/setCategory [Status: 200]
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] User-Agent: Medusa/0.2.2 (Windows; 10; cdb05a61-5ab4-11e8-92f1-7cdd9083bde4)
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] With post data: category=medusa&hashes=2b44e208ac311fae598013251bb7bfbd9e16ba9d
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] qbittorrent: Response to POST request is 
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] qbittorrent: Requested a POST connection to http://localhost:9496/command/increasePrio with params: None Data: {u'hashes': '2b44e208ac311fae598013251bb7bfbd9e16ba9d'}
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] POST URL: http://localhost:9496/command/increasePrio [Status: 200]
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] User-Agent: Medusa/0.2.2 (Windows; 10; cdb05a61-5ab4-11e8-92f1-7cdd9083bde4)
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] With post data: hashes=2b44e208ac311fae598013251bb7bfbd9e16ba9d
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] qbittorrent: Response to POST request is 
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] Checking if we should stop searching for a better quality for for episode Black.Clover.-.S01E32.MKV./.H.264./.AAC./.HDTV./.Legendado./.Japonês./.1080p
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] Matched release G:\Meu Drive\Media\Animes\Black Clover\Season 1\Black.Clover.S01E32.032.1080p.HDTV-Multiple Subtitle to a series in your database: Black Clover
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] Parsed G:\Meu Drive\Media\Animes\Black Clover\Season 1\Black.Clover.S01E32.032.1080p.HDTV-Multiple Subtitle into title: Black Clover, season: 1, episode: [32], episode_title: 032, screen_size: 1080p, format: HDTV, release_group: Multiple Subtitle, type: episode, parsing_time: 0.169000148773, total_time: 0.169000148773, absolute_episode: [], quality: 1080p HDTV
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] Found codec for Black Clover Three-Leaf Sprouts
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] 331753: Episode has no release group, replacing it with Multiple Subtitle
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] 331753: Formatting pattern: %SN - %Sx32 - 032 - %EN - %QN -> Black Clover - 1x32 - 032 - Three-Leaf Sprouts - 1080p HDTV
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] Sending a Telegram message for Black Clover - 1x32 - 032 - Three-Leaf Sprouts - 1080p HDTV with 8 seeders and 0 leechers from BJ-Share
2018-05-18 13:04:07 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] Telegram in use with API KEY: **********
2018-05-18 13:04:08 INFO     SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] Telegram message sent successfully.
2018-05-18 13:04:08 DEBUG    SNATCHQUEUE-MANUALSNATCH-331753 :: [71edd70] Transaction with 1 queries executed```

Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Tested it without slashes, and same happens, it will not send to qbittorrent (but if i remove the accents, it send just fine, so it must be something with accents or encoding)...
Tested it to blackhole, and it work fine even with accents, so must be something with the api call to qbittorrent...

Comment (duramato) [reply medusa/providers/torrent/html/bjshare.py]: Ok, please open an issue for that then. That's an issue with qBittorrent and affects all providers with non ascii results. For testing provider for now, you can use blackhole.

Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Ok, i'll open an issue, i did an big refactor to the code, if you can take a look if it's fine now..
(i'll need to do more tests to be sure that everything is working, i did just few tests)

**Comment by DarkSupremo** [comment]
The travis is failing due to an error in outside the scope of this pull request:

/home/travis/build/pymedusa/Medusa/medusa/search/core.py:87:43: N806 variable 'fileOut' in function should be lowercase

I already did an merge with develop, still happens

**Comment by p0psicles** [comment]
@DarkSupremo if you do another merge from develop, it should build now.

File: medusa/providers/torrent/html/bjshare.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        # Proper Strings
+        self.proper_strings = ["PROPER", "REPACK", "REAL", "RERIP"]
+
+        # Cache
+        self.cache = tv.Cache(self, min_time=30)
+
+        self.quality = OrderedDict([
+            ('Full HD ', '1080p'),
+            ('HD ', '720p'),
+            ('4K', '2160p')])
```
Comment (duramato) [inline medusa/providers/torrent/html/bjshare.py]: `4k`Isn't needed, it  gets parsed correctly so it's fine as is,
```
$ guessit "4K"
For: 4K
GuessIt found: {
    "screen_size": "4K",
    "type": "movie"
}
```

File: medusa/providers/torrent/html/bjshare.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    """BJ-Share Torrent provider."""
+
+    def __init__(self):
+        """Initialize the class."""
+        super(BJShareProvider, self).__init__('BJ-Share')
+
+        # URLs
+        self.url = 'https://bj-share.me'
+        self.urls = {
+            'detail': urljoin(self.url, "torrents.php?id=%s"),
```
Comment (duramato) [inline medusa/providers/torrent/html/bjshare.py]: both `detail` and `download` are unused? Remove them

File: medusa/providers/torrent/html/bjshare.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+            for result in torrent_rows[1:]:
+                cells = result("td")
+                if len(cells) < len(labels):
+                    continue
+                try:
+                    title = cells[labels.index("Nome/Ano")].find("a", dir="ltr").get_text(strip=True)
+                    # get international title if available
+                    title = re.sub(".* \[(.*?)\](.*)", r"\1\2", title)
+
+                    for serie in self.absolute_numbering:
```
Comment (duramato) [inline medusa/providers/torrent/html/bjshare.py]: This shouldn't be a thing. Either make it work for all animes in absulute numbering or just set the show to not anime in Medusa. I'm more inclined to the latter.

Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: One piece is the only anime that i'm aware that is in "absolute" numbering, the problem is that they include the season (wrong season) and episode as absolute, eg: One Piece - S08E836
836 is the latest episode in absolute numbering, that is correct, but S08 is not the current season...
So for this show, i don't see a other way to make it work, setting it as not anime will not fix it i think...

Comment (p0psicles) [reply medusa/providers/torrent/html/bjshare.py]: So this indexer is consistently indexing the wrong season, for one piece releases? So there are no other release groups available for one piece, which are indexed correctly?.

Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: So this indexer is consistently indexing the wrong season, for one piece releases?
Yes, i already complained about that and seens like they don't care to fixing it (and many other broken things)...

So there are no other release groups available for one piece, which are indexed correctly?
You mean in another indexer on in this one? 

In this indexer, it looks that it is added "automatically", so all current and new releases will be broken until they or the source from where they get that info fix it...

And currently there is not another brazillian indexer in Medusa, i have plans to add more when i have free time (B2S-Share, Manicomio Share and Speed-Share)

Comment (p0psicles) [reply medusa/providers/torrent/html/bjshare.py]: ok, in that case I don't mind fixing it in this provider as an exception.
Please include in the code the reason why this exception (technical debt) is there. So that in the future, we can easily remove it, when they decide to fix it.

Comment (p0psicles) [reply medusa/providers/torrent/html/bjshare.py]: Wait, you only noticed this for One piece, or multiple animes?

Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: @p0psicles only for One piece

Comment (medariox) [reply medusa/providers/torrent/html/bjshare.py]: This is a guessit issue and should not be fixed on the provider side.

Comment (medariox) [reply medusa/providers/torrent/html/bjshare.py]: If that is the case, okay.

File: medusa/providers/torrent/html/bjshare.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+        self.max_back_pages = 2
+        self.units = ["B", "KB", "MB", "GB", "TB", "PB"]
+
+        # Proper Strings
+        self.proper_strings = ["PROPER", "REPACK", "REAL", "RERIP"]
+
+        # Cache
+        self.cache = tv.Cache(self, min_time=30)
+
+        self.quality = OrderedDict([
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: Why is this an OrderedDict and not just a dict?

Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Otherwise Full HD could be parsed to Full 720p if the check for HD happens first...

Comment (medariox) [reply medusa/providers/torrent/html/bjshare.py]: Remove this and just use replace.

File: medusa/providers/torrent/html/bjshare.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+        self.required_cookies = ['session']
+
+        # Torrent Stats
+        self.minseed = None
+        self.minleech = None
+
+        # Miscellaneous Options
+        self.supports_absolute_numbering = True
+        self.max_back_pages = 2
+        self.units = ["B", "KB", "MB", "GB", "TB", "PB"]
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: Default units not needed.

File: medusa/providers/torrent/html/bjshare.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+                    if not response:
+                        log.debug("No data returned from provider")
+                        continue
+
+                    result = self._parse(response.content, mode, search_string)
+                    has_next_page = result['has_next_page']
+                    items += result['items']
+
+                # For each search mode sort all the items by seeders if available
+                items.sort(key=lambda d: try_int(d.get('seeders', 0)), reverse=True)
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: Needs to go as well.

File: medusa/providers/torrent/html/bjshare.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+        items = []
+        has_next_page = False
+        with BS4Parser(data, "html5lib") as html:
+            torrent_table = html.find("table", id="torrent_table")
+            torrent_rows = torrent_table("tr") if torrent_table else []
+
+            # year_re = re.search(r"\((\d{4})\)|(\d{4})", search_string, re.I)
+
+            # ignore next page in RSS mode
+            has_next_page = mode != 'RSS' and html.find("a", attrs={"class", "pager_next"}) is not None
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: Use `class_=`

File: medusa/providers/torrent/html/bjshare.py [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+                    torrent_name = "{0} {1}".format(title, torrent_details.strip()).strip()
+                    torrent_name = re.sub('\s+', ' ', torrent_name)
+
+                    items.append({
+                        'title': torrent_name,
+                        'link': download_url,
+                        'size': size,
+                        'seeders': seeders,
+                        'leechers': leechers,
+                        'hash': ''
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: Remove `hash`, add `pubdate`.

Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: This tracker does not include pubdate in the torrent listing page...
hash removed

File: medusa/providers/torrent/html/bjshare.py (hunk 10 above)
Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Done, it was copied from Alpharatio, should be removed there in future also so...

File: medusa/providers/torrent/html/bjshare.py (hunk 9 above)
Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Done, it was copied from Alpharatio, should be removed there in future also so...

File: medusa/providers/torrent/html/bjshare.py (hunk 8 above)
Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Ok, done.
Did this way because there could be more cases that i did not saw initially..

File: medusa/providers/torrent/html/bjshare.py (hunk 11 above)
Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Done

**Comment by DarkSupremo** [comment]
Added support to grouping torrent settings, i'll do more tests to make sure everything is working

File: medusa/providers/torrent/html/bjshare.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+        # URLs
+        self.url = 'https://bj-share.me'
+        self.urls = {
+            'search': urljoin(self.url, "torrents.php")
+        }
+
+        # Credentials
+        self.enable_cookies = True
+        self.cookies = ''
+        self.required_cookies = ['session']
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: Why did you add this if the provider doesn't require captcha? Why is using user and pass not working?

File: medusa/providers/torrent/html/bjshare.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+
+                    download_url = urljoin(self.url, result.select('a[href^="torrents.php?action=download"]')[0]['href'])
+                    if not all([title, download_url]):
+                        continue
+
+                    seeders = try_int(cells[labels.index('Seeders')+group_index].get_text(strip=True))
+                    leechers = try_int(cells[labels.index('Leechers')+group_index].get_text(strip=True))
+
+                    # Filter unseeded torrent
+                    if seeders < self.minseed or leechers < self.minleech:
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: See AlphaRatio on how this should look like.

File: medusa/providers/torrent/html/bjshare.py [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+            log.debug(u'More Pages? {0}'.format(has_next_page))
+
+            # Continue only if at least one Release is found
+            if len(torrent_rows) < 2:
+                log.debug('Data returned from provider does not contain any torrents')
+                return {'has_next_page': has_next_page, 'items': []}
+
+            # '', '', 'Name /Year', 'Files', 'Time', 'Size', 'Snatches', 'Seeders', 'Leechers'
+            labels = [process_column_header(label) for label in torrent_rows[0]('td')]
+            group_title = u''
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: The `u` prefix needs to go.

File: medusa/providers/torrent/html/bjshare.py [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+
+            # Skip column headers
+            for result in torrent_rows[1:]:
+                cells = result('td')
+                result_class = result.get("class")
+                group_index = -2 if "group_torrent" in result_class else 0
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: I still see a lot of double quotes.

File: medusa/providers/torrent/html/bjshare.py [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+
+                    torrent_name = '{0} {1}'.format(title, torrent_details.strip()).strip()
+                    torrent_name = re.sub('\s+', ' ', torrent_name)
+
+                    items.append({
+                        'title': torrent_name,
+                        'link': download_url,
+                        'size': size,
+                        'seeders': seeders,
+                        'leechers': leechers,
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: Add pubdate: None

File: medusa/providers/torrent/html/bjshare.py [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+                    })
+
+                    if mode != 'RSS':
+                        log.debug('Found result: {0} with {1} seeders and {2} leechers'.format
+                                  (torrent_name, seeders, leechers))
+
+                except StandardError:
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: Don't use StandardError. See other providers

File: medusa/providers/torrent/html/bjshare.py [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+                    continue
+        return {'has_next_page': has_next_page, 'items': items}
+
+    def login(self):
+        """Login method used for logging in before doing a search and torrent downloads."""
+        return self.cookie_login('Login', check_url=self.urls['search'])
```
Comment (medariox) [inline medusa/providers/torrent/html/bjshare.py]: Use default login methods if there is no captcha.

File: medusa/providers/torrent/html/bjshare.py (hunk 13 above)
Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: It does require captcha... it's just not always but something trigger it from time to time, i'm not sure what exacly, i had problem with that already, that's why i did with cookies...

File: medusa/providers/torrent/html/bjshare.py (hunk 14 above)
Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: What exacly?
That's almost identical to AlphaRatio, the difference is that it's addapted to support Grouping Torrents settings, alpharatio does not support that...

When it's enabled, the layout is a bit different, also the columns, that's why i need the group_index that fix it based on if grouping torrents is enabled or not

File: medusa/providers/torrent/html/bjshare.py (hunk 15 above)
Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: I'm not python developer at daily but that field needs to be unicode, there's no problem if i init it as str and change to unicode later?

File: medusa/providers/torrent/html/bjshare.py (hunk 16 above)
Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Yeah, java developer habits... changing it

File: medusa/providers/torrent/html/bjshare.py (hunk 15 above)
Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Done, changed to None

File: medusa/providers/torrent/html/bjshare.py (hunk 17 above)
Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: Done

File: medusa/providers/torrent/html/bjshare.py (hunk 19 above)
Comment (DarkSupremo) [reply medusa/providers/torrent/html/bjshare.py]: It does require captcha... it's just not always but something trigger it from time to time, i'm not sure what exacly, i had problem with that already, that's why i did with cookies...

**Comment by DarkSupremo** [comment]
Unless someone find something else to change, for me, the development of this provider is finished, already did all planned for this...

**Comment by DarkSupremo** [comment]
@fernandog yes i have, send me your email
